synfig-core/src/modules/lyr_freetype/lyr_freetype.cpp

```diff
@@ -264,6 +252,99 @@ class FaceCache {
 	}
 };
 
+/**
+ * Map font filenames to their FreeType faces
+ */
+struct FaceMap
+{
+	typedef std::map<filesystem::Path, FT_Face> MapType;
+	typedef MapType::const_iterator const_iterator;
+	typedef MapType::iterator iterator;
+
+	FaceMap() = default;
+
+	const_iterator
+	begin() const
+	{
+		return map_.begin();
+	}
+
+	const_iterator
+	end() const
+	{
+		return map_.end();
+	}
+
+	const_iterator
+	find(const filesystem::Path& path) const
+	{
+		return map_.find(path);
+	}
+
+	FT_Face&
+	operator[](const filesystem::Path& path)
+	{
+		return map_[path];
+	}
+
+	void clear()
+	{
+		for (const auto& item : map_)
+			FT_Done_Face(item.second);
+	}
+
+	~FaceMap()
+	{
+		clear();
+	}
+
+	FaceMap(const FaceMap&) = delete; // Copy prohibited
+	void operator=(const FaceMap&) = delete; // Assignment prohibited
+	FaceMap(FaceMap&&) = delete; // Move constructor prohibited
+	FaceMap& operator=(FaceMap&&) = delete; // Move assignment prohibited
+
+private:
+	std::map<filesystem::Path, FT_Face> map_;
+};
+
+/**
+ * Metadata to be stored in FT_Face->generic field
+ */
+struct FaceMetaData
+{
+	filesystem::Path path;
+#if HAVE_HARFBUZZ
+	hb_font_t* font{nullptr};
+#endif
+
+	static FaceMetaData&
+	get_from_face(FT_Face face)
+	{
+		return *static_cast<FaceMetaData*>(face->generic.data);
+	}
+
+	void
+	add_to_face(FT_Face face)
+	{
+		if (face->generic.data)
+			face->generic.finalizer(face);
+		face->generic.data = this;
+		face->generic.finalizer = FaceMetaData::self_destroy;
+	}
+
+private:
+	static void
+	self_destroy(void* object)
+	{
+		FT_Face face = static_cast<FT_Face>(object);
+		FaceMetaData* meta_data = static_cast<FaceMetaData*>(face->generic.data);
+		hb_font_destroy(meta_data->font);
+		delete meta_data;
```

Comment:
Don't you need to clear `face->generic.data` here?
I mean:
```
face->generic.data = nullptr;
```

Reply:
I'll try it

Reply:
It seems it worked :D rebasing it